dataservice/api/biospecimen/models.py

```diff
@@ -111,3 +131,114 @@ class Biospecimen(db.Model, Base):
     biospecimen_genomic_files = db.relationship(BiospecimenGenomicFile,
                                                 backref='biospecimen',
                                                 cascade='all, delete-orphan')
+    diagnoses = db.relationship('Diagnosis', secondary='biospecimen_diagnosis',
+                                backref=db.backref(
+                                    'biospecimens'))
+
+
+def validate_diagnosis_biospecimen(target):
+    """
+    Ensure that both the diagnosis and biospecimen
+    have the same participant
+    If this is not the case then raise DatabaseValidationError
+    """
+    from dataservice.api.errors import DatabaseValidationError
+    # Return if biospecimen_diagnosis is None
+    if not target:
+        return
+
+    # Get biospecimen and diagnosis by id
+    bsp = None
+    ds = None
+    if target.biospecimen_id and target.diagnosis_id:
+        bsp = Biospecimen.query.get(target.biospecimen_id)
+        ds = Diagnosis.query.get(target.diagnosis_id)
+
+    # If biospecimen and diagnosis doesn't exist, return and
+    # let ORM handle non-existent foreign key
+    if bsp is None or ds is None:
+        return
+
+    # Check if this diagnosis and biospecimen refer to same participant
+    if ds.participant_id != bsp.participant_id:
+        operation = 'modify'
+        target_entity = BiospecimenDiagnosis.__tablename__
+        message = (
+            ('a diagnosis cannot be linked with a biospecimen if they '
+             'refer to different participants. diagnosis {} '
+             'refers to participant {} and '
+             'biospecimen {} refers to participant {}')
+            .format(ds.kf_id,
+                    ds.participant_id,
+                    bsp.kf_id,
+                    bsp.participant_id))
+        raise DatabaseValidationError(target_entity, operation, message)
+
+
+def validate_biospecimen(target):
+    """
+    Ensure that both the diagnosis and biospecimen
+    have the same participant
+    If this is not the case then raise DatabaseValidationError
+    """
+    from dataservice.api.errors import DatabaseValidationError
+    # Return if biospecimen is None
+    if not target:
+        return
+    print(target.diagnoses)
+    # Get diagnosis by id
+    ds = None
+    if target.diagnoses:
```

Comment:
I think these two lines can be removed and line 186 can become `if not target or not target.diagnoses:`

Reply:
👍